src/middleware/ember_html.rs

```diff
@@ -26,11 +67,63 @@ pub async fn serve_html(request: Request, next: Next) -> Response {
+            return (StatusCode::METHOD_NOT_ALLOWED, headers).into_response();
+        }
+
+        // Come up with an Open Graph image URL. In case a crate page is requested,
+        // we use the crate's name and the OG image base URL from config to
+        // generate one, otherwise we use the fallback image.
+        let og_image_url = 'og: {
+            if let Some(suffix) = path.strip_prefix(PATH_PREFIX_CRATES) {
+                let len = suffix.find('/').unwrap_or(suffix.len());
+                let krate = &suffix[..len];
+
+                // `state.config.og_image_base_url` will always be `Some` as that's required
+                // if `state.config.index_html_template_path` is `Some`, and otherwise this
+                // middleware won't be executed; see `crate::middleware::apply_axum_middleware`.
+                if let Ok(og_img_url) = state.config.og_image_base_url.as_ref().unwrap().join(krate)
+                {
+                    break 'og Cow::from(og_img_url.to_string());
+                }
+            }
+            OG_IMAGE_FALLBACK_URL.into()
+        };
+
+        // Fetch the HTML from cache given `og_image_url` as key or render it
+        let html = RENDERED_HTML_CACHE
+            .get_or_init(|| init_html_cache(state.config.html_render_cache_max_capacity))
+            .get_with_by_ref(&og_image_url, async {
+                // `OnceLock::get_or_init` blocks as long as its intializer is running in another thread.
+                // Note that this won't take long, as the constructed Futures are not awaited
+                // during initialization.
+                let template_env = TEMPLATE_ENV.get_or_init(|| {
+                    // At this point we can safely assume `state.config.index_html_template_path` is `Some`,
+                    // as this middleware won't be executed otherwise; see `crate::middleware::apply_axum_middleware`.
+                    init_template_env(state.config.index_html_template_path.clone().unwrap())
+                        .boxed()
+                        .shared()
+                });
+
+                // Render the HTML given the OG image URL
+                let env = template_env.clone().await;
+                let html = env
+                    .get_template(INDEX_TEMPLATE_NAME)
+                    .unwrap()
+                    .render(context! { og_image_url})
+                    .expect("Error rendering index");
+
+                html
+            })
+            .await;
+
         // Serve static Ember page to bootstrap the frontend
-        ServeFile::new("dist/index.html")
-            .oneshot(request)
-            .await
-            .map(|response| response.map(axum::body::Body::new))
+        Response::builder()
+            .header(header::CONTENT_TYPE, "text/html")
+            .header(header::CONTENT_LENGTH, html.len())
+            .body(axum::body::Body::new(html))
```

Comment:
I guess we could probably use https://docs.rs/axum/latest/axum/response/struct.Html.html for this?